docs/design/factories.md

```diff
@@ -0,0 +1,234 @@
+provider-controlled][structs-interfaces].
+
+This document is meant to catalogue our available options and explore their
+benefits and trade-offs.
+
+## Options
+
+### Anonymous functions
+
+Wherever we need these types from provider developers, we can ask for them as
+the results of anonymous functions. For example:
+
+```go
+type Provider struct {
+  Resources map[string]func() framework.Resource
+}
+```
+
+This is asking for a function that returns a `framework.Resource` when called,
+which lets the framework call the function to instantiate a new instance of the
+type. This works whether the type is defined by the provider (the function
+would return an interface) or by the framework.
+
+### Named functions
+
+Similarly to anonymous functions, we can ask for named functions:
+
+```go
+type ResourceFactory func() framework.Resource
+
+type Provider struct {
+  Resources map[string]ResourceFactory
+}
+```
+
+This also asks for a function that returns a `framework.Resource`, but gives
+the function signature a name. This works whether the type is defined by the
+provider (the function would return an interface) or by the framework.
+
+### Reflection
+
+In theory, given a single instance of the type, we can use the `reflect`
+package to create new instances of the type:
+
+
+```go
+resource := computeInstanceResource{}
+
+typ := reflect.TypeOf(resource)
+newResource := reflect.New(typ)
+
+// newResource is now a newly instantiated variable of the same type as
+// resource
+```
+
+We can use this in a helper, to do something like:
+
+```go
+type Provider struct {
+  Resources map[string]Resource
+}
+```
+
+and do the instantiation ourselves behind the scenes.
+
+This works if the type is defined by the provider, but there's not really any
+point to it if the type is defined by the framework, as the framework can just
+use manual copying to get the same result.
+
+### Manual copying
+
+If the Resource type is owned by the framework, it will be able to instantiate
+its own instances of that type, and would be able to copy over the data used to
+populate them:
+
+```go
+resource := framework.Resource{
+	Schema: map[string]*Schema{},
+	Create: createFunc,
+	Read: readFunc,
+	Update: updateFunc,
+	Delete: deleteFunc,
+}
+
+newResource := framework.Resource{}
+newResource.Schema = resource.Schema
+newResource.Create = resource.Create
+newResource.Read = resource.Read
+newResource.Update = resource.Update
+newResource.Delete = resource.Delete
+```
+
+we can use this in a helper, to do something like:
+
+```go
+type Provider struct {
+	Resources map[string]Resource
+}
+```
+
+and do the instantiation ourselves behind the scenes.
+
+This works only if the type is defined by the framework. Types defined by the
+provider would need to use reflection to achieve this outcome.
+
+### Factory types
+
+Instead of using functions, we can have an interface for a factory type that
+providers define:
```

Comment:
```suggestion
can be implemented by a provider-defined type:
```

We might prefer "fulfilled" to "implemented" here - I just noticed that the Go language spec never uses "fulfil".